### DISCUSSION TITLE
Authentication for private Package not working

### DISCUSSION THREAD

**Discussion by dovidgef** [renovatebot/renovate#10543] [Request Help] [answered]
I am attempting to configure Renovate to automatically authenticate for NPM package stored in private Github NPM Repository.

But I keep getting 404 error in the logs. I see that despite my config below Renovate is still using the default NPM registry for this package `https://registry.npmjs.org` instead of the configured registry `https://npm.pkg.github.com`.

Any ideas why the below config is not working?

Thanks

<br>Debug Log:
```
   "message": "Response code 404 (Not Found)",
    "stack": "HTTPError: Response code 404 (Not Found)\n    at Request.<anonymous> (/home/ubuntu/renovateapp/node_modules/renovate/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)",
    "options": {
      "headers": {
        "user-agent": "Renovate Bot (GitHub App 2740)",
        "accept": "application/json",
        "accept-encoding": "gzip, deflate, br"
      },
      "url": "https://registry.npmjs.org/@organizationName%2Fnpm-toolkit",
      "username": "",
      "password": "",
      "method": "GET",
      "http2": false
    },
    "response": {
      "statusCode": 404,
      "statusMessage": "Not Found",
      "body": {
        "error": "Not found"
      },
      "headers": {
        "date": "Tue, 22 Jun 2021 16:24:03 GMT",
        "content-type": "application/json",
        "content-length": "21",
        "connection": "close",
        "cf-request-id": "0ad62247cd000039eafcb44000000001",
        "expect-ct": "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"",
        "vary": "Accept-Encoding",
        "server": "cloudflare",
        "cf-ray": "6636d31fad9d39ea-SEA"
      },
      "httpVersion": "1.1"
    }
  },
  "packageName": "@organizationName/npm-toolkit"
}
```

<br>Renovate Config:
```
{
  "extends": ["config:base", ":prConcurrentLimitNone", ":prHourlyLimitNone"],
  "hostRules": [
    {
      "baseUrl": "https://npm.pkg.github.com",
      "encrypted": {
        "token": "<Encrypted PAT Token"
      }
    }
  ],
  "packageRules": [
    {
      "matchUpdateTypes": ["minor", "patch"],
      "matchCurrentVersion": "!/^0/",
      "automerge": true
    },
    {
      "matchPackagePrefixes": ["@organizationName"],
      "datasources": ["npm"],
      "registryUrls": ["https://npm.pkg.github.com"]
    }
  ],
}
```

**Comment by dovidgef** [comment]
I just notice this gem in my logs:

```
WARN: Custom registries are not allowed for this datasource and will be ignored
{
  "datasource": "npm",
  "registryUrls": [
    "https://npm.pkg.github.com"
  ]
}
```

Any idea why this is, and how to get around it?

**Reply by viceice** [reply]
You need to configure a `.npmrc` file in repo root or configure `npmrc` renovate option, but this will fail on lockfile update. As we run npm / yarn / pnpm to update lockfile.

**Reply by dovidgef** [reply]
I thought when you set the `npmrc` option in the config, Renovate automatically generates a `.npmrc` file for you.
So why will updating lockfile fail in this case?

**Reply by viceice** [reply]
OK, it seems we write the `npmrc` config to `.npmrc` before running package manage to update lockfile. so you can use one of the two possibillities.

**Comment by dovidgef** [comment]
Btw, I am using the Renovate's Github Application in this case.

Bottom line, is there any way for me to configure Renovate to fully support a private NPM package hosted on Github's private NPM repository?

**Reply by viceice** [reply]
yes

**Reply by dovidgef** [reply]
Can you point me to a sample configuration that would work in this case?

I appreciate your help.

**Reply by viceice** [reply]
It depends on the manager you are using. yarn / npm / pnpm, which one are you using?

**Reply by dovidgef** [reply]
Using npm

**Reply by dovidgef** [reply]
This config also needs to apply only to this specific private package all other packages need to still use standard registry without any auth.

**Reply by viceice** [reply]
So it's a scoped package?

**Reply by viceice** [reply]
How do you configure npm to access the private package when you work locally / github actions ci ?

**Reply by dovidgef** [reply]
Yes its scoped.

**Reply by dovidgef** [reply]
This is how our Github actions workflow looks:

```
- name: Setup Node
        uses: actions/setup-node@v1
        with:
          node-version: ${{ env.NODE_VER }}
          registry-url: https://npm.pkg.github.com
          scope: '@organizationName'

      - name: Install npm packages
        uses: bahmutov/npm-install@v1
        env:
          NPM_FLAGS: --production=false
          NODE_AUTH_TOKEN: ${{ secrets.GH_PAT }}
```

**Reply by viceice** [reply]
if you have a repo `.npmrc` like this:

```ini
@scope:registry=https://npm.pkg.github.com/<org-name>/
```

it should work with following config:

```json
{
  "hostRules": [
    {
      "matchHost": "npm.pkg.github.com",
      "hostType": "npm",
      "encrypted": {
        "token": "<Encrypted PAT Token>"
      }
    }
  ]
}
```

**Reply by dovidgef** [reply]
Ok thanks, I will try that.

**Reply by viceice** [reply]
I would suggest to use a repo `.npmrc` over renovate `npmrc` config. but this should also work:

```json
{
  "hostRules": [
    {
      "matchHost": "https://npm.pkg.github.com/",
      "hostType": "npm",
      "encrypted": {
        "token": "<Encrypted PAT Token>"
      }
    }
  ],
  "npmrc": "@organizationName:registry=https://npm.pkg.github.com/"
}
```

I've often seen issues when the trailing slash is missing on registry urls.

**Reply by viceice** [reply]
maybe you also need `//npm.pkg.github.com/:always-auth=true` to force auth

**Reply by viceice** [reply]
can you please post the working config as new answer, so we can add it to our docs as sample for others.

**Reply by dovidgef** [reply]
```
{
  "hostRules": [
    {
      "matchHost": "npm.pkg.github.com",
      "hostType": "npm",
      "encrypted": {
        "token": "<Encrypted PAT Token>"
      }
    }
  ]
}
```
This config is still not working, still giving 404 trying to find private package on `https://registry.npmjs.org`

I will try your second alternative config now.

**Comment by dovidgef** [comment, marked as answer]
This config did the trick:

```
{
  "hostRules": [
    {
      "matchHost": "https://npm.pkg.github.com/",
      "hostType": "npm",
      "encrypted": {
        "token": "<Encrypted PAT Token>"
      }
    }
  ],
  "npmrc": "@organizationName:registry=https://npm.pkg.github.com/"
}
```

Thanks for your help

**Reply by dovidgef** [reply]
I did not test a lockfile update yet, I hope to test it soon.

**Reply by dovidgef** [reply]
I now tested updating the lockfile and that works as well with this config.

**Reply by viceice** [reply]
@HonkingGoose Can we add this to our docs for private npm modules please?

**Reply by HonkingGoose** [reply]
See PR #10574. 😉

**Reply by ruben-chainalysis** [reply]
[This section of the docs](https://docs.renovatebot.com/getting-started/private-packages/#add-an-encrypted-npm-token-to-renovate-config) uses `npmToken` instead of just `token` inside the `encrypted` block, adding the `NPM_TOKEN` then to the `npmrc`. That didn't work for my use case, using the hosted app. Following @dovidgef's example did. For Artifactory in particular, a block like this does the trick (you can scope the registry for an organization as @dovidgef or overwrite them all like I did, in case you want to check all dependencies in Artifactory, where you might have a virtual registry mirroring public ones):

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "hostRules": [
    {
      "matchHost": "https://yourdomain.here/artifactory/api/npm/npm/",
      "hostType": "npm",
      "encrypted": {
        "token": "somethingEncryptedWithTheHostedAppKey"
      }
    }
  ],
  "npmrc": "registry=https://yourdomain.here/artifactory/api/npm/npm/\nalways-auth=true\nemail=the_token_owners@fakeemailaccount.io"
}
```

This made Renovate work when looking up npm and Yarn 1.x dependencies. I hope it saves someone else as much time as I spent banging my head against the wall.
